### DISCUSSION TITLE
unable to generate sourcemap for sass with asset/resource

### DISCUSSION THREAD

**Discussion by samuelcolvin** [webpack/webpack#13505] [Q&A] [answered]
Thank you for webpack, it's awesome.

My setup is successfully compiling sass/scss and outputting a `.css` file which I'm referencing in js, however unfortunately I can't get a `.css.map` (source map for the scss) file to be generated too.

What I'd really like is the equivalent of

    ./node_modules/.bin/sass --source-map-urls=relative src/styles/main.scss dist/assets/main.css

which generates `main.css` and `main.css.map`.

I have the following module rules in `webpack.config.js`:

```js
    rules: [
      {
        test: /\.tsx?$/,
        loader: 'ts-loader',
        options: {
          // transpileOnly: true,
        },
      },
      {
        test: /\.s[ac]ss$/i,
        type: 'asset/resource',
        generator: {
          filename: 'assets/[name].css[query]',
        },
        use: [
          {
            loader: 'sass-loader',
            options: {
              sourceMap: true,
              sassOptions: {
                outputStyle: 'expanded',
              },
            },
          },
        ],
      },
      {
        test: /\.(css|png|ico|jpe?g|svg|woff2?|ttf)$/i,
        type: 'asset/resource',
      },
    ],
```

I I know that the `sass-loader` is generating a source map - I inserted a log statement here:

https://github.com/webpack-contrib/sass-loader/blob/6ceaa0fe8bce480159d06e856680f2edab13847a/src/index.js#L83

and the map is present and populated. However `webpack` seems to be discarding the map while I need it to be written to `assets/[name].css.map`.

I've tried numerous values of `devtool` but it doesn't seem to make any difference. I've also tried adding `style-loader` and `css-loader`, but that seems to break the css generating and doesn't cause a source map to be generated.

Is this a limitation of `type: 'asset/resource'`, or am I doing something wrong? Could I implement my own loader to accomplish this?

Notable dependency version:

```json
    "css-loader": "^5.2.6",
    "sass": "^1.34.1",
    "sass-loader": "^12.0.0",
    "style-loader": "^2.0.0",
    "ts-loader": "^9.2.2",
    "webpack": "^5.38.1",
    "webpack-cli": "^4.7.0"
```

**Comment by alexander-akait** [comment, marked as answer]
Please pass these options https://sass-lang.com/documentation/js-api#sourcemap to `sassOptions`, not `sourceMap`, the `sourceMap` option for internal source maps (handled and extracted/inlined by webpack), but here you need external source maps (i.e. from `sass`), but I do not recommend do it

**Reply by samuelcolvin** [reply]
You mean something like:

```js
      {
        test: /\.s[ac]ss$/i,
        type: 'asset/resource',
        use: [
          {
            loader: 'sass-loader',
            options: {
              sassOptions: {
                sourceMap: true,  // tried as path too
                sourceMapEmbed: true,
                outputStyle: 'expanded',
              },
            },
          },
        ],
      },
```
?

I've tried that but still:
* no reference to the sourcemap in the css file
* no sourcemap file generated

Perhaps that means this is a problem with `sass-loader`?

**Reply by alexander-akait** [reply]
hm, I think yes, can you open an issue in `sass-loader` with small repo example (so I can test)

**Reply by samuelcolvin** [reply]
Sorry, I'm super busy and needed a solution asap. I can create an issue, but I don't have the time to create a demo repo.

For anyone else struggling with this, I created my own simple loader as follows:

```js
const sass = require('sass')
const path = require('path')

module.exports = async function () {
  const {resourcePath, emitFile, async} = this
  const callback = async()
  const output_name = path.parse(resourcePath).base.replace(/\.s[ac]ss$/, '.css')
  sass.render({
    file: resourcePath,
    outFile: output_name,
    sourceMap: true,
    sourceMapContents: true,
    sourceMapRoot: '/styles/',
  }, (err, result) => {
    if (err) {
      callback(err)
    } else {
      // hide my local development path from prying eyes.
      const map = result.map.toString().replace(/file:\/\/.+?\/([^\/]+\.s[ac]ss)/g, '$1')
      // output path of the map file is currently (somewhat) hard coded
      emitFile(`assets/${output_name}.map`, map)
      callback(null, result.css)
    }
  })
}
```
(**edit:** cleaned up script a little)

usage:

```js
      {
        test: /\.s[ac]ss$/i,
        type: 'asset/resource',
        generator: {
          filename: 'assets/[name].[hash].css[query]',
        },
        loader: path.join(__dirname, 'custom-sass-loader.js')
      },
```

**Reply by alexander-akait** [reply]
I strongly recommend do not do it

**Reply by samuelcolvin** [reply]
I'd much prefer not to do it this way, but I needed to get something to work, and this is working.

I'm deploying to cloudflare workers, not standard browser, thus I need a separate `.css.map` file, not a `.js.map` file.

**Reply by alexander-akait** [reply]
You lose all dep (i.e. `@import`/`@use`/etc) and lose tracking them, it will be big problem, not only for watch, also for caches

**Reply by alexander-akait** [reply]
You can click `...` on your issue and create the issue from original question, then format it

**Reply by samuelcolvin** [reply]
https://github.com/webpack-contrib/sass-loader/issues/962
